### PR TITLE
GODRIVER-1827 Implement a pausable connection pool.

### PR DESCRIPTION
**Pull request by matthewdale**
[GODRIVER-1827](https://jira.mongodb.org/browse/GODRIVER-1827)

Update the connection pool to use the "paused", "ready", and "closed" state transitions and corresponding actions. 

Behavior changes:
* Update the `pool` state transitions to use "paused", "ready", and "closed" as described in the updated CMAP spec.
* Update a `Server` to call `pool.ready()` whenever the `Server` state changes to anything except "Unknown".
* Move responsibility for calling SDAM connection handshake error handling from `connection.connect()` into the callers of `connection.connect()`
  * Allows delivering the original connection error to callers of `pool.checkOut()` _before_ the SDAM handshake error handler calls `pool.clear()`, which delivers a different error to all goroutines waiting on `pool.checkOut()`.
  * Allows preventing the RTT monitor from updating server state (SDAM spec does not describe that the RTT monitor should use SDAM error handling rules).
* Make the error returned when the pool is cleared retryable.
  * Refactor `Operation.Execute()` retry logic to allow making connection check-out retryable.
* Add constants for `ConnectionCheckOutStartedEvent` and `PoolReadyEvent`.

Code improvements:
* Simplify `connection.connect()` API by returning connection errors directly from `connect()` instead of requiring callers to call `wait()` immediately after to get the error.
* Send `ConnectionCheckOutStartedEvent` in `pool.checkOut()` instead of `Server.Connection()`

Test changes:
* Sync all non-integration CMAP spec tests.
* Update CMAP spec test runner to work with new "ready" command.
* Update unified spec test runner to support `ConnectionCheckOutStartedEvent` and `PoolReadyEvent`.

### CODE REVIEW

**Review by kevinAlbs — COMMENTED** [review]
I reviewed compared the spec changes from https://github.com/mongodb/specifications/commit/beda39620fb39938a8d46eeaf63b75f807fe7c4e and https://github.com/mongodb/specifications/commit/b4423500b018352227b7df8d21b36681db0ad58a to review major changes in this PR. I will review line-by-line after we meet. I do not have any significant feedback or open questions. So far, this looks solid!

The [Connection Pool Management SDAM prose test](https://github.com/mongodb/specifications/commit/beda39620fb39938a8d46eeaf63b75f807fe7c4e#diff-023626308692bf0090aadd07b8ac54a169e1d7e7c6bdc621d23c97c848081b4cR213) may need to be added, unless I missed it.

File: x/mongo/driver/topology/pool.go
Comment (kevinAlbs) [inline x/mongo/driver/topology/pool.go:903]: This came up during our meeting. If the pool is cleared just after L900, `maintain` may try to rebuild the pool. It seems unlikely to occur. The `maintainInterval` is 10 seconds and not configurable by public API.

I think the worst case is when `minPoolSize` is large and `maxConnecting` is large. For example, suppose `minPoolSize` is 1000 and `maxConnecting` is 100:
- 1000 `wantConn` are added to `newConnWait`.
- Each of the 100 `createConnections` creates a connection.
   - If any fail to connect, the pool will be cleared.
   - If all succeed at connecting. They will be checked in and unused until the server state is updated to something aside from "Unknown".

The impact does not seem too harmful to me. Consider adding a comment to describe this possible scenario.

**Review by benjirewis — COMMENTED** [review]
Excellent job so far; this is pretty dense.

**Review by kevinAlbs — COMMENTED** [review]
I appreciate the clarifying comments. I have a few open questions, but no major feedback. Fantastic work!

File: x/mongo/driver/topology/server_test.go
Comment (matthewdale) [reply x/mongo/driver/topology/server_test.go:422]: In the old behavior, the only sequence possible is:
1. Set `returnConnectionError = true`.
2. `server.Connection()` -> `pool.checkOut()` requests a connection via a `wantConn`.
3. `pool.createConnections()` picks up the `wantConn` and calls `conn.connect()`.
4. `conn.connect()` encounters an error and calls `server.ProcessHandshakeError()`, which clears the pool and increments the generation number.
4. `pool.createConnections()` sends the error to the `wantConn`.
5. `server.Connection()` returns the error.
6. Repeat from step 2.

There are no pool state changes other than incrementing the generation number, so every call to `server.Connection()` after setting `returnConnectionError = true` causes a connection error, clears the pool, and increments the generation number. 

In the new behavior, the following sequence is possible:
1. Set `returnConnectionError = true`.
2. `server.Connection()` -> `pool.checkOut()` requests a connection via a `wantConn`.
3. `pool.createConnections()` picks up the `wantConn` and calls `conn.connect()`, which returns an error.
4. `pool.createConnections()` sends the error to the `wantConn`.
5. `server.Connection()` returns the error.
6. `server.Connection()` -> `pool.checkOut()` requests a connection via a `wantConn`.
7. `pool.createConnections()` calls `server.ProcessHandshakeError()` which clears the pool, sets it to "paused", and increments the generation number. All subsequent calls to `pool.checkOut()` will now return a "pool paused" error.
7. `pool.createConnections()` picks up the `wantConn` and calls `conn.connect()`, which returns an error.
8. `pool.createConnections()` sends the error to the `wantConn`.
9. `pool.createConnections()` calls `server.ProcessHandshakeError()` which clears the pool and increments the generation number.
10. `server.Connection()` -> `pool.checkOut()` returns a "pool paused" error and does not increment the generation number.
11. Repeat from step 10.

The pool transitions to a "paused" state when it is cleared and we should expect fewer pool generation number increments because checking out from a "paused" pool doesn't increment the generation number. However, because `server.Connection()` can return an error before `pool.createConnections()` calls `server.ProcessHandshakeError()`, the pool won't necessarily be "paused" by the next time `server.Connection()` is called in the `for` loop at L438. The result is that up to 2 pool clears can happen before the pool is "paused".

Note: I'm actually confused about why the data race detector doesn't consider updating `returnConnectionError` a data race. `returnConnectionError` is read from a different goroutine than the test case goroutine (from the `pool.createConnections()` goroutine) without anything obvious synchronizing read/write access.

Comment (kevinAlbs) [reply x/mongo/driver/topology/server_test.go:422]: There is one part I still do not understand. Wouldn't the call to `server.ProcessHandshakeError` in step 7 also clear `newConnWait`? 

As I understand it. When a pool enters the "paused" state, any current `checkOut` calls may add a `wantConn` to the queue after the pool enters the "paused" state.

File: event/monitoring.go
Comment (matthewdale) [reply event/monitoring.go:78]: I'm not sure why the existing pool event constants are named different than the strings, but changing them would be a backward-breaking change since they're exported and not in the `x/` directory. I decided to stick with the existing pattern for consistency, even tho the pattern is weird.

File: x/mongo/driver/operation.go
Comment (matthewdale) [reply x/mongo/driver/operation.go:347]: Based on @benjirewis's comment about the possible relevance to CSOT, I'll leave it and add the suggested test. Thanks for the detailed code!

File: x/mongo/driver/topology/CMAP_spec_test.go
Comment (matthewdale) [reply x/mongo/driver/topology/CMAP_spec_test.go:130]: Ah, good find. AFAIK a negative value is not currently used because passing a negative duration to `time.NewTicker()` will cause a panic.

Do you think it's worth adding the logic to handle negative values now (i.e. call `WithConnectionPoolMaintainInterval` with a very large value) or just leave it as-is and handle the test panic when it comes up?

File: x/mongo/driver/topology/pool.go
Comment (matthewdale) [reply x/mongo/driver/topology/pool.go:655]: Added a check that doesn't clear the pool if the current state is "closed".

File: x/mongo/driver/operation.go
Comment (matthewdale) [reply x/mongo/driver/operation.go:352]: Added test `TestRetry` that includes a test that asserts that retry mode `RetryContext` works as expected.

File: x/mongo/driver/topology/pool.go
Comment (matthewdale) [reply x/mongo/driver/topology/pool.go:923]: Actually I realized I solved the same race condition in `checkOut()` by holding the `stateMu` read lock when entering the wait queue, which prevents state changes between checking the state and entering the wait queue. I've updated `maintain()` to hold the `stateMu` read lock while modifying the wait queue as well.

File: x/mongo/driver/topology/server.go
Comment (matthewdale) [reply x/mongo/driver/topology/server.go:228]: Added a comment here explaining why we start pools in "ready" when we `Connect()` a server.

File: x/mongo/driver/topology/CMAP_spec_test.go
Comment (matthewdale) [reply x/mongo/driver/topology/CMAP_spec_test.go:130]: Updated to disable the `maintain()` goroutine if given a negative `MaintainInterval`.

File: x/mongo/driver/topology/pool.go
Comment (benjirewis) [reply x/mongo/driver/topology/pool.go:261]: Gotcha, sounds good, thanks.

File: x/mongo/driver/topology/connection_test.go
Comment (benjirewis) [reply x/mongo/driver/topology/connection_test.go:78]: Sounds good; thanks for the explanation here and offline.

File: event/monitoring.go
Comment (benjirewis) [reply event/monitoring.go:78]: Sounds good.

**Review by benjirewis — APPROVED** [review]
LGTM mod outstanding open questions. Fantastic work @matthewdale 😄

File: x/mongo/driver/topology/server_test.go
Comment (kevinAlbs) [inline x/mongo/driver/topology/server_test.go:216]: Am I correct that this test may have failed before if `close(eventsCh)` ran before the "ConnectionPoolCleared" event was emitted? If so, nice find. The new approach seems more resilient.

File: x/mongo/driver/topology/pool.go
Comment (kevinAlbs) [reply x/mongo/driver/topology/pool.go:923]: Oh nice, that is good to it is capped at 10 new connections.

File: x/mongo/driver/topology/server.go
Comment (kevinAlbs) [reply x/mongo/driver/topology/server.go:228]: Thank you for adding the comment. I think this can be resolved from changes in 351bf9ee724df5d658b2af7355a6046e32560421.

File: x/mongo/driver/operation.go
Comment (kevinAlbs) [reply x/mongo/driver/operation.go]: I think this can resolved given the new `RetryablePoolError` interface in 351bf9ee724df5d658b2af7355a6046e32560421.

File: x/mongo/driver/topology/CMAP_spec_test.go
Comment (kevinAlbs) [reply x/mongo/driver/topology/CMAP_spec_test.go:130]: I think this is resolved from the changes in 586f967c98656f564f04d9c843e3940da916bd94.

**Review by kevinAlbs — APPROVED** [review]
LGTM. Investigating intermittent test failures is no easy feat. Well done!

I have two open questions, but nothing to block merging.

File: x/mongo/driver/operation.go
Comment (divjotarora) [inline x/mongo/driver/operation.go:545]: Given that this exact code is duplicated in the `WriteCommandError` and `Error` cases, can we write a closure within this function to handle this logic? Suggesting a closure rather than a separate helper function as it'll be easier to do all of the state management like setting `srvr` and `conn` to nil in a closure that can directly access those variables.

**Review by divjotarora — APPROVED** [review]
Looks great! Per our discussion, I focused my review on `operation.go`. Everything looks good and I just have a couple of minor comments, but no need to re-request review if the changes aren't too big. LGTM!

File: x/mongo/driver/topology/server_test.go
Comment (matthewdale) [reply x/mongo/driver/topology/server_test.go:216]: `pool.close()` (called by `server.Disconnect()`) waits for all background goroutines to exit before returning, so it shouldn't be possible for events to be sent after `server.Disconnect()` returns. The problem is actually that `pool.clear()` was called after `pool.close()` in the "operation context timeout with unrelated dial errors should clear the pool" test case. After responding to @benjirewis's feedback, I updated `pool.clear()` to be a no-op if the pool is "closed", so the "ConnectionPoolCleared" event was no longer sent and the test failed. The updated test waits for the "ConnectionPoolCleared" event to be sent before calling `server.Disconnect()`, resolving that issue.

File: x/mongo/driver/operation.go
Comment (matthewdale) [reply x/mongo/driver/operation.go:545]: Sounds like a good idea, will add a closure that resets the variable states for retry.

File: x/mongo/driver/operation.go
Comment (matthewdale) [reply x/mongo/driver/operation.go:375]: I don't think the `Connection` interface nor any of the implementations document that `Close()` is idempotent. However, the existing implementation does both `defer conn.Close()` and `conn.Close()` on retry, so the behavior here matches previous behavior and it seems to suggest that all `conn.Close()` implementations used here are idempotent. I think it's worth documenting the expected behavior of `Close()` and the other functions in the `Connection` interface, but I'm not sure adding additional logic to handle different implementations of `Close()` is necessary.

Comment (matthewdale) [reply x/mongo/driver/operation.go:382]: Added a comment that describes the expected behavior of `Connection.Close()` implementations.

File: x/mongo/driver/topology/server_test.go
Comment (matthewdale) [reply x/mongo/driver/topology/server_test.go:429]: Ok, I think I finally got to the bottom of the issue. My previous description is impossible in the latest code because of the `pool.stateMu` synchronization that prevents pool state changes while modifying the wait queue.

The actual root cause of the previous test failures is that `pool.checkOut()`/`Server.Connection()` returns an error before the `createConnections()` goroutine handles the error, including removing the connection from the pool. As a result, it's possible for the the generation connection count to be off immediately after `Server.Connection()` returns until the `createConnections()` goroutine handles the connection error. The fix is to wait up to a short amount of time for the connection count to reach the target value. Note that this is the same fix that I had to make to the `TestPool/checkOut/handshaker_i/o_fails` test.

The comment above `WithMaxConnecting` is still correct and we should still set `maxConnecting=1` to make the test reliable. I've updated the test code to use `assert.Eventually()` to wait up to 100ms for the generation connection count to reach the target value and made the comparison strict equality, not an greater/less comparison.

File: x/mongo/driver/operation.go
Comment (matthewdale) [reply x/mongo/driver/operation.go]: Added the closure `resetForRetry` that captures most of the common code for resetting the retry loop variables.

File: x/mongo/driver/topology/server_test.go
Comment (kevinAlbs) [reply x/mongo/driver/topology/server_test.go:216]: Thank you for the explanation.

**Pull request merged**
